### PR TITLE
8333755: NumberFormat integer only parsing breaks when format has suffix

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR and [CSR](https://bugs.openjdk.org/browse/JDK-8333920) which corrects a bug where NumberFormat cannot successfully parse values in integer only mode, when the format expects a suffix.

For example,
```
// a format that expects a currency suffix
var fmt = NumberFormat.getCurrencyInstance(Locale.FRANCE);
fmt.setParseIntegerOnly(true);
failFmt.parse("5,00 €"); // throws ParseException when you would have expected 5 returned
```

When parsing in integer only mode, instead of breaking upon a decimal symbol encounter, we should store the index but continue to fully parse so that we can verify the entire string and increment our position to search for and match the suffix. Upon a successful suffix match, we can then set `ParsePosition.index` back to the stored decimal index.

It should be noted that CompactNumberFormat did previously have code that would traverse the rest of the String, to allow matching of the suffix. The difference is that NumberFormat returns the index upon decimal symbol encounter, while CompactNumberFormat was implemented to return the index reflected by the entire string traversal. Such differences cannot be standardized due to behavioral compatibility concerns.

Thus while parsing in integer only, CNF sets index to the `Position.fullPos()`, while DF sets index to the `Position.intPos()`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8333920](https://bugs.openjdk.org/browse/JDK-8333920) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8333755](https://bugs.openjdk.org/browse/JDK-8333755): NumberFormat integer only parsing breaks when format has suffix (**Bug** - P4)
 * [JDK-8333920](https://bugs.openjdk.org/browse/JDK-8333920): NumberFormat integer only parsing breaks when format has suffix (**CSR**)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19664/head:pull/19664` \
`$ git checkout pull/19664`

Update a local copy of the PR: \
`$ git checkout pull/19664` \
`$ git pull https://git.openjdk.org/jdk.git pull/19664/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19664`

View PR using the GUI difftool: \
`$ git pr show -t 19664`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19664.diff">https://git.openjdk.org/jdk/pull/19664.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19664#issuecomment-2161279197)

### CODE REVIEW

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 1]
```diff
@@ -2437,6 +2441,9 @@ private final boolean subparse(String text, ParsePosition parsePosition,
         return true;
     }
 
+    // To track the full parse index as well as the integer portion
+    record Position(int intPos, int fullPos) {}
+
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/DecimalFormat.java:2446]: The record name may be more descriptive. Also I would expect more description here, such as what it is and those params are.

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 2]
```diff
@@ -2452,11 +2459,12 @@ private final boolean subparse(String text, ParsePosition parsePosition,
      * @return returns the position of the first unparseable character or
      *         -1 in case of no valid number parsed
      */
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/DecimalFormat.java:2461]: Needs modification accommodating `Position`.

File: test/jdk/java/text/Format/NumberFormat/StrictParseTest.java [hunk 3]
```diff
@@ -279,7 +313,13 @@ public void parsePositionIndexTest() {
     // ---- Helper test methods ----
 
     // Should parse entire String successfully, and return correctly parsed value.
-    private double successParse(NumberFormat fmt, String toParse) {
+    private Number successParse(NumberFormat fmt, String toParse) {
+        return successParse(fmt, toParse, toParse.length());
+    }
+
+    // Overloaded method used for integer only parsing, where the expected
```
Comment (naotoj) [inline test/jdk/java/text/Format/NumberFormat/StrictParseTest.java:320]: Is this method for `integer only` only?

**Review by naotoj — COMMENTED** [review]
Looks good overall. Some minor nits follow.

File: test/jdk/java/text/Format/NumberFormat/StrictParseTest.java (hunk 3 above)
Comment (justin-curtis-lu) [reply test/jdk/java/text/Format/NumberFormat/StrictParseTest.java]: A leftover comment, now removed.

File: src/java.base/share/classes/java/text/DecimalFormat.java (hunk 1 above)
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/text/DecimalFormat.java]: Thanks for the review Naoto.

Renamed to `NumericPosition`, as well as added further descritpion. Hesitant to rename to anything longer/detail too much, for what is essentially a limited scope helper class.

**Review by naotoj — APPROVED** [review]
LGTM

**Pull request closed without merging**